cli/cmd/component-delete.go

```diff
@@ -0,0 +1,185 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"helm.sh/helm/v3/pkg/action"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/kinvolk/lokomotive/pkg/components"
+	"github.com/kinvolk/lokomotive/pkg/components/util"
+	"github.com/kinvolk/lokomotive/pkg/config"
+	"github.com/kinvolk/lokomotive/pkg/k8sutil"
+)
+
+var componentDeleteCmd = &cobra.Command{
+	Use:   "delete",
+	Short: "Delete an installed component",
+	Long: `Delete a component.
+When run with no arguments, all components listed in the configuration are deleted.`,
+	Run: runDelete,
+}
+
+var deleteNamespace bool
+
+// nolint:gochecknoinits
+func init() {
+	componentCmd.AddCommand(componentDeleteCmd)
+	pf := componentDeleteCmd.PersistentFlags()
+	pf.BoolVarP(&deleteNamespace, "delete-namespace", "", false, "Delete namespace with component.")
+}
+
+func runDelete(cmd *cobra.Command, args []string) {
+	contextLogger := log.WithFields(log.Fields{
+		"command": "lokoctl component delete",
+		"args":    args,
+	})
+
+	lokoCfg, diags := getLokoConfig()
+	if len(diags) > 0 {
+		contextLogger.Fatal(diags)
+	}
+
+	componentsToDelete := make([]string, len(args))
+	copy(componentsToDelete, args)
+
+	// Verify if all the components user has specified on command line exist.
+	for _, componentName := range componentsToDelete {
+		if _, err := components.Get(componentName); err != nil {
+			contextLogger.Fatal(err)
+		}
+	}
+
+	if len(args) == 0 {
+		componentsToDelete = make([]string, len(lokoCfg.RootConfig.Components))
+
+		// We don't need to verify if these components exist because this is already verified when
+		// the config file is loaded.
+		for i, component := range lokoCfg.RootConfig.Components {
+			componentsToDelete[i] = component.Name
+		}
+	}
+
+	if !askForConfirmation(
+		fmt.Sprintf(
+			"The following components will be deleted:\n\t%s\n\nAre you sure you want to proceed?",
+			strings.Join(componentsToDelete, "\n\t"),
+		),
+	) {
+		contextLogger.Info("Components deletion cancelled.")
+		return
+	}
+
+	kubeconfig, err := getKubeconfig()
+	if err != nil {
+		contextLogger.Fatalf("Error in finding kubeconfig file: %s", err)
+	}
+
+	if err := deleteComponents(lokoCfg, kubeconfig, componentsToDelete...); err != nil {
+		contextLogger.Fatal(err)
+	}
+}
+
+func deleteComponents(lokoCfg *config.Config, kubeconfig string, componentNames ...string) error {
+	for _, componentName := range componentNames {
+		fmt.Printf("Deleting component '%s'...\n", componentName)
+
+		// This is already verified, so we can ignore the error.
+		component, _ := components.Get(componentName)
```

Comment:
In verifying loop in main function, we could collect those objects, rather than their names, so we don't have to collect them twice.

Reply:
fixed.

Reply:
This last commit also fixes https://github.com/kinvolk/lokomotive/issues/266